src/dataflow/src/server.rs

```diff
@@ -917,44 +919,47 @@ impl PendingPeek {
             cursor.step_key(&storage);
         }
 
+        // Cursor and bound lifetime for `Row` data in the backing trace.
         let (mut cursor, storage) = self.trace_bundle.oks_mut().cursor();
+        // Accumulated `Vec<Datum>` results that we are likely to return.
         let mut results = Vec::new();
+        // Temporary storage for datums we are considering.
+        let mut datums = Vec::new();
 
-        // We can limit the record enumeration if i. there is a limit set,
-        // and ii. if the specified ordering is empty (specifies no order).
-        let limit = if self.finishing.order_by.is_empty() {
-            self.finishing.limit.map(|l| l + self.finishing.offset)
-        } else {
-            None
-        };
+        let mut row_packer = repr::RowPacker::new();
+        let arena = RowArena::new();
+
+        // There are roughly two paths we can take here, where we split
+        // based on whether there is an ordering or not. If there is not
+        // an ordering, we power through all the elements perhaps bailing
+        // early if a limit is reached. We hit each row with any operators
+        // for post-processing (map, filter, project).
+        // If there is an ordering, we want to more carefully pilot things
+        // to leave the data as `[Datum]` and use this representation to
+        // maintain the top so-many results. We would love to use a heap
+        // to maintain this efficiently, but Rust makes this tricky.
```

Comment:
Why wouldn't we be able to use a BTreeMap as the heap?

Reply:
Afaik, it will only sort things by their `Ord` implementation, whereas we have a custom ordering method.